packages/node/src/indexer/worker/block-dispatcher.service.ts

```diff
@@ -0,0 +1,389 @@
+import { NodeConfig } from '../../configure/NodeConfig';
+import { AutoQueue, Queue } from '../../utils/autoQueue';
+import { getLogger } from '../../utils/logger';
+import { fetchBlocksBatches } from '../../utils/substrate';
+import { ApiService } from '../api.service';
+import { IndexerEvent } from '../events';
+import { IndexerManager } from '../indexer.manager';
+import {
+  FetchBlock,
+  ProcessBlock,
+  InitWorker,
+  NumFetchedBlocks,
+  NumFetchingBlocks,
+  SetCurrentRuntimeVersion,
+} from './worker';
+import { Worker } from './worker.builder';
+
+type IIndexerWorker = {
+  processBlock: ProcessBlock;
+  fetchBlock: FetchBlock;
+  numFetchedBlocks: NumFetchedBlocks;
+  numFetchingBlocks: NumFetchingBlocks;
+  setCurrentRuntimeVersion: SetCurrentRuntimeVersion;
+};
+
+type IInitIndexerWorker = IIndexerWorker & {
+  initWorker: InitWorker;
+};
+
+type IndexerWorker = IIndexerWorker & {
+  terminate: () => Promise<number>;
+};
+
+async function createIndexerWorker(): Promise<IndexerWorker> {
+  const indexerWorker = Worker.create<IInitIndexerWorker>(
+    path.resolve(__dirname, '../../../dist/indexer/worker/worker.js'),
+    [
+      'initWorker',
+      'processBlock',
+      'fetchBlock',
+      'numFetchedBlocks',
+      'numFetchingBlocks',
+      'setCurrentRuntimeVersion',
+    ],
+  );
+
+  await indexerWorker.initWorker();
+
+  return indexerWorker;
+}
+
+type GetRuntimeVersion = (block: SubstrateBlock) => Promise<RuntimeVersion>;
+
+function getMaxWorkers(numWorkers?: number): number {
+  const maxCPUs = os.cpus().length;
+  return Math.min(numWorkers ?? maxCPUs, maxCPUs);
+}
+
+export interface IBlockDispatcher {
+  init(
+    runtimeVersionGetter: GetRuntimeVersion,
+    onDynamicDsCreated: (height: number) => Promise<void>,
+  ): Promise<void>;
+
+  enqueueBlocks(heights: number[]): void;
+
+  queueSize: number;
+  freeSize: number;
+  latestBufferedHeight: number | undefined;
+
+  // Remove all enqueued blocks, used when a dynamic ds is created
+  flushQueue(height: number): void;
+}
+
+const logger = getLogger('BlockDispatcherService');
+
+// TODO move to another file
+/**
+ * @description Intended to behave the same as WorkerBlockDispatcherService but doesn't use worker threads or any parallel processing
+ */
+@Injectable()
+export class BlockDispatcherService
+  implements IBlockDispatcher, OnApplicationShutdown
+{
+  private fetchQueue: Queue<number>;
+  private processQueue: AutoQueue<void>;
+
+  private fetching = false;
+  private isShutdown = false;
+  private getRuntimeVersion: GetRuntimeVersion;
+  private onDynamicDsCreated: (height: number) => Promise<void>;
+  private _latestBufferedHeight: number;
+
+  constructor(
+    private apiService: ApiService,
+    private nodeConfig: NodeConfig,
+    private indexerManager: IndexerManager,
+    private eventEmitter: EventEmitter2,
+  ) {
+    this.fetchQueue = new Queue(nodeConfig.batchSize * 3);
+    this.processQueue = new AutoQueue(nodeConfig.batchSize * 3);
```

Comment:
cucurrency is 1 then, I suppose
```suggestion
    this.processQueue = new AutoQueue(nodeConfig.batchSize * 3, nodeConfig.batchSize);
```

Reply:
if batchSize is overall batchSize, then concurrency will become `nodeConfig.batchSize / workers`